src/ol/resolutionconstraint.js

```diff
@@ -2,37 +2,76 @@
  * @module ol/resolutionconstraint
  */
 import {linearFindNearest} from './array.js';
-import {clamp} from './math.js';
+import {getHeight, getWidth} from './extent';
+import {clamp} from './math';
 
 
 /**
- * @typedef {function((number|undefined), number, number): (number|undefined)} Type
+ * @typedef {function((number|undefined), number, import("./size.js").Size, boolean=): (number|undefined)} Type
  */
 
+/**
+ * @param {number} resolution Resolution
+ * @param {import("./extent.js").Extent=} maxExtent Maximum allowed extent.
+ * @param {import("./size.js").Size} viewportSize Viewport size.
+ * @return {number} Capped resolution.
+ */
+function getCappedResolution(resolution, maxExtent, viewportSize) {
+  const xResolution = getWidth(maxExtent) / viewportSize[0];
+  const yResolution = getHeight(maxExtent) / viewportSize[1];
+  return Math.min(resolution, Math.min(xResolution, yResolution));
+}
+
+/**
+ * @param {number} resolution Resolution.
+ * @param {number} maxResolution Max resolution.
+ * @param {number} minResolution Min resolution.
+ * @return {number} Smoothed resolution.
+ */
+function getSmoothedResolution(resolution, maxResolution, minResolution) {
```

Comment:
What does this function do? I'm failing to understand.

Reply:
This applies a "smooth" clamp, meaning the resolution is allowed to be slightly out of bounds. I'll clarify it